File: app/src/main/java/com/rowland/delivery/features/dash/presentation/viewmodels/SharedViewModel.java
package com.rowland.delivery.features.dash.presentation.viewmodels;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import java.util.List;

/**
 * Created by Rowland on 5/9/2018.
 */

public abstract class SharedViewModel<T> extends ViewModel {

    protected final MutableLiveData<List<T>> dataList = new MutableLiveData<>();
    protected final MutableLiveData<T> selectedListItem = new MutableLiveData<>();

    public void setSelectedListItem(int position) {
        selectedListItem.setValue(dataList.getValue().get(position));
    }

    public LiveData<T> getSelectedListItem() {
        return selectedListItem;
    }

    public abstract LiveData<List<T>> getDataList();
}

File: app/src/main/java/com/rowland/delivery/features/dash/presentation/viewmodels/SharedViewModel.kt
package com.rowland.delivery.features.dash.presentation.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.rowland.delivery.features.dash.presentation.data.Resource

/**
 * Created by Rowland on 5/9/2018.
 */

abstract class SharedViewModel<T> : StateViewModel() {

    protected val dataList = MutableLiveData<Resource<List<T>>>()
    protected val selectedListItem = MutableLiveData<T>()

    open fun setSelectedListItem(position: Int) {
        selectedListItem.value = dataList.value!!.data!![position]
    }

    fun getSelectedListItem(): LiveData<T> {
        return selectedListItem
    }

    fun getDataList(): LiveData<Resource<List<T>>> {
        return dataList
    }
}